### ISSUE TITLE
Question: Broadcasting in Gaussian Processes

### ISSUE THREAD

**Issue by mileslucas** [tensorflow/probability#248]
Hello,

I am new to TFP and I'm running into issues trying to fit a Gaussian Process. I can't seem to get the broadcasting correct.

For background, I have a problem where there are `n` index points where each point has 3 dimensions. My index points look like
```
X.shape = [264, 3]
```
My data points have a number of outputs independent from the number of kernel dimensions, so my observations look like
```
Y.shape = [264, 4]
```

In defining my kernel, I use an RBF kernel with one length scale for each dimension in my index points
```
amplitude = tf.Variable(10., name='amplitude')
length_scales = tf.Variable([300, 30, 30], name='length_scales')
kernel = tfp.positive_semidefinite_kernels.ExponentiatedQuadratic(
    amplitude,
    length_scales,
)
```
This kernel has batch shape of 3

Now, my Gaussian Process is defined using the kernel and the index points
```
gp = tfd.TransformedDistribution(
    tfd.GaussianProcess(
        kernel=kernel
        index_points = X
    ), 
    bijector = tfp.bijectors.Reshape(event_shape_out=[Y.shape[-1]]
)
```
This reshape creates the mapping `3 -> 4` that I need to go from my 3 input dimensions to my 4 output dimensions.

Now, when I try and get the log likelihood with my `Y` I get bad errors
```
nll = -gp.log_prob(Y)
```
So, my Gaussian Process should take in 264 events with 3 inputs each. The kernel should use these 3 inputs for getting the covariance matrix. Then, the GP should give me back 4 outputs. However, the above fails saying dimensions 264 and 4 are not compatible. 

I have played around with a lot of the settings and can't seem to figure out the dimensionality of the problem. Sorry for such a noob question, but how can I successfully train my `[n, F]` inputs to `[n, K]` observations for, and eventually use the model as a map from `F -> K`?

**Comment by axch** [comment]
Correct me if I'm wrong, but it looks to me like you're trying to train a batch of 4 3-input Gaussian Processes (one GP per dimension in each Y vector).  If you want them to share kernel parameters, then you should just get rid of the reshape bijector, and put the batch dimension in Y on the left.  That is,
```none
gp = tfd.GaussianProcess(
        kernel=kernel
        index_points=X)

Y_with_batch_on_top = tf.reshape(Y, ...)
# Y_with_batch_on_top.shape = [4, 264]
```
Then `gp.log_prob` should give you a batch of 4 log probability values, which you can sum or whatever.

If you want the 4 GPs to have different kernel parameters, you can add a batch dimension to your Variables too:
```none
amplitude = tf.Variable([10., 11., 12., 13.] name='amplitude')
length_scales = tf.Variable(
    [[300, 30, 30], [301, 31, 31], [302, 32, 32], [303, 33, 33]], name='length_scales')
```

**Comment by mileslucas** [comment]
Hi, thanks for this I'll try it out.

To clarify-
what I need is a 3-dimensional kernel, such that the kernel takes 3 parameter inputs and returns 1 output.


![image](https://user-images.githubusercontent.com/14099459/49980281-701c4900-ff18-11e8-9bb0-9cc322196e31.png)

I think these are called ARD (automatic relevance determination) kernels.

IE- the kernel should be defined as batch shape 4 with feature shape 3 such that if I called `kern.matrix(X, X)` I would get a `4 x len(X) x len(X)` tensor. I've found a way to do this by simply multiplying three kernels together.

**Comment by axch** [comment]
Sounds like we can close this issue as resolved, then.  Feel free to reopen, or start a new issue, if you run into any more trouble.

**Comment by mileslucas** [comment]
So there does not exist a kernel like the one I described? Do I have to multiply many kernels together instead of using the batch shaping?

**Comment by csuter** [comment]
Hi @mileslucas, indeed we don't currently have an ARD implementation. We *should*, we just haven't checked one in yet :) -- please see below for more on how to do this if you want to unblock while we work on checking in a proper implementation.

However I think there some other conceptual issues going on here, which I'll try to clarify. Forgive the impending lack of brevity :).

You have 264 inputs in a 3-d space. Let's give these values some names. In TFP GP/kernel context, we call that 264 the number of examples, or "the examples dim(ension)". The 3 we call the number of features, or "the features dim". When we instantiate a GP with `E` examples, the result represents the marginal distribution over function values at those `E` input locations (264 here) -- this is effectively an `E`-dimensional multivariate normal distribution. In TFP, an `E`-dimensional multivariate distribution is said to have "event shape `[E]`", i.e. the shape of an element of the sample space (an event) is `[E]`. So in a GP, the number of examples, `E`, *becomes* the event shape of the resulting distribution. Now, you have for each input location 4 separate output values. I'm guessing you *don't* want to treat all of these as being drawn from the same GP (maybe you do -- in that case things are a bit simpler; let me know if I've got this wrong). In this case you actually want to work with 4 independent GP instances. In TFP we call this a "batch", or we say that the (desired) GP distribution instance has "batch shape [4]". Assuming you want to fit separate kernel parameters for each of these GPs (since, e.g., the 4 outputs each vary with different characteristic length scales), you can accomplish this by creating a kernel of batch shape 4. This really means you have 4 *separate* kernels, but we "batch" them together because it allows an efficient vectorization of the requisite computations. (You said above that your kernel has batch shape 3 -- this is because you tried to create 3 length scales (one per feature dim) but this isn't currently supported in the basic kernels we've implemented -- again, more on ARD below). Here's an example of how this could look so far:

```python
# sample 264 random input locations in 3d unit cube
x_ = np.random.uniform(size=[264, 3])
# sample 264 batches of 4 output values. note that i've ordered these differently than you have them.
y_ = np.random.uniform(size=[4, 264])

amplitudes = tf.Variable(np.random.uniform(4), "amplitudes")
# *NOT* ARD yet, just a batch of 4 independent values 
length_scales = tf.Variable(np.random.uniform(4), "length_scale")

kernel = tfp.positive_semidefinite_kernels.ExponentiatedQuadratic(amplitudes, length_scales)
print(kernel.batch_shape)
#  ==> [4]

gp = tfp.distributions.GaussianProcess(
    kernel=kernel,
    index_points=x_)
print(gp.batch_shape)
#  ==> [4]
print(gp.event_shape)
# ==> [264]

print(gp.sample(20).shape)
# ==> [20, 4, 264]

# Now we can take the log probs of our observations -- we get 4 results, which
# represent the 4 individual log probs of the observations according to the respective
# GP models in the batch.
print(gp.log_prob(y_).shape)
# ==> [4]

# Now, if you want to consider this as a "joint" model over the 4 outputs, we can use
# the Independent meta-distribution to "reinterpret" batch dims as event dims.
joint_gp = tfp.distributions.Independent(gp, reinterpreted_batch_ndims=1)
print(joint_gp.batch_shape)
# ==> []
print(joint_gp.event_shape)
# ==> [4, 264]

# And this will give a scalar result.
print(joint_gp.log_prob(y_).shape)
# ==> []
```

Note we can't use the Reshape bijector here. Bijectors only operate on event shapes. If you want to transform (independent) batch shapes into (joint) event shapes, you can wrap with the `Independent` meta-distribution. Note, you could also skip using Independent and do the reduce_sum that @axch mentioned; the effect is the same. For a great summary of shape semantics in TFP, see @derifatives's fantastic [tutorial on TFP shapes](https://colab.research.google.com/github/tensorflow/probability/blob/master/tensorflow_probability/examples/jupyter_notebooks/Understanding_TensorFlow_Distributions_Shapes.ipynb). 

Now, ARD is a popular (and quite sensible) choice, which we haven't actually implemented! I was actually just talking to some colleagues about this this week, and have a sketch of an implementation that I'll be formalizing and checking in shortly. Until then, you can check it out, below, and feel free to try it out yourself. Please let me know if it doesn't work (I'll be finding out soon enough :))!

As to why you won't be able to get ARD simply by using a batch of length scales, this has to do with the fact that we have to cleverly inject `1`s in the shapes of the inputs to compute the `kernel.matrix(...)` output (see the implementation in the abstract base class `PositiveSemidefiniteKernel`. To do ARD (*and* support independent batching of kernel params, which we also need), we'd need to inject 1's in between the rightmost (feature) dims and the left-most (batch) dims. I believe the right way to think of this is that the ARD params are actually parameterizing a transformation of the input space *before* applying the kernel, and in fact the implementation I am planning looks exactly like this. I was actually just working on it earlier this week, and the rough sketch looks thus (in case you'd like to copy/paste it locally until we have it checked in properly): 

(I haven't tested this code yet, but I *think* it should work as written)
```python
class InputTransformedKernel(PositiveSemidefiniteKernel):

  def __init__(self, kernel, transformation):
    self._kernel = kernel
    self._transformation = transformation

  def _apply(self, x1, x2, param_expansion_ndims=0):
    return kernel._apply(
        self._transformation(x1),
        self._transformation(x2),
        param_expansion_ndims)


class InputScaledKernel(InputTransformedKernel):

  def __init__(self, kernel, length_scales):
    super(InputScaledKernel, self).__init__(
        kernel,
        lambda x: x / tf.expand_dims(length_scales,
                                     -(kernel.feature_ndims + 1))
```

Note the 1 we have to inject into the shape there. That is so that the "examples" dim of the inputs (the one that represents the multiple inputs over which we are computing the kernel matrix) will be broadcast across.

The calling code would then look like:

```python
length_scales = tf.Variable(np.random.uniform(size=[4, 3], "length_scales")
se_kernel = tfk.ExponentiatedQuadratic(amplitude)  # length_scale = None here, implicitly

# This is the "ARD" kernel (we don't like abbreviations or bizarrely obscure names in
# TFP, so we're probably going to call this "InputScaledKernel" since....that's what it is! :)
kernel = InputScaledKernel(se_kernel, length_scales)

# The batch shape is still the same -- the 3 per-feature length scales will be
# "eaten" by the kernel calculation.
print(kernel.batch_shape)
# ==> [4]

gp = tfd.GaussianProcess(kernel, index_points)
```

Another workaround, as you alluded to above, would be to multiply the separate kernels together. This, as I'm sure you realize, is totally clunky and won't be as efficient (since it's not able to be fully vectorized).

I hope this helps to clarify things! Sorry for the wall of text; shape issues are always the trickiest part of dealing with flowing `Tensor`s.  Please hit me back with any questions you have. Hopefully this screed contains adequate information to get you moving again.

And **thanks** for using TFP and trying out our GP functionality!

**Comment by csuter** [comment]
(I'll reopen until you've got things working)

**Comment by mileslucas** [comment]
This really helps clear up not only how TFP broadcasting works but also helps clarify my own problem. The code I'm trying to rewrite is for loops on for loops and I am really excited to try and use the benefits of having many dimension tensors and operating on them without for loops. I don't like for loops! 

As for the implementation above, I run into issues with subclassing the PositiveSemidefiniteKernel

```python
class InputTransformedKernel(tfk.PositiveSemidefiniteKernel):

    def __init__(self, kernel, transformation, feature_ndims=1, name='InputTransformedKernel'):
        super().__init__(feature_ndims, name=name)
        self._kernel = kernel
        self._transformation = transformation

    def _apply(self, x1, x2, param_expansion_ndims=0):
        return kernel._apply(
            self._transformation(x1),
            self._transformation(x2),
            param_expansion_ndims)

class InputScaledKernel(InputTransformedKernel):
    
    def __init__(self, kernel, length_scales):
        trans = lambda x: x / tf.expand_dims(length_scales,
                                     -(kernel.feature_ndims + 1))
        super().__init__(kernel,trans)
```

First off, when I naively run
```python
se_kernel = tfk.ExponentiatedQuadratic(amplitude=a)
kernel = InputScaledKernel(se_kernel, len_scales)
sigs = kernel.matrix(gparams, gparams)
```
where `a` has shape `(4, 1)` and  `gparams` has shape `(264, 3)`. I expect for `sigs` to be a `(4, 264, 264)` tensor (in this specific example), but I get errors-

```
InvalidArgumentError: Dimensions must be equal, but are 3 and 4 for 'InputTransformedKernel_3/InputTransformedKernel/truediv' (op: 'RealDiv') with input shapes: [3,1,264], [4,1,3].
```

In addition, the batch_shape private method needs to be written for `InputTransformedKernel`, as well as any other abstract methods that I don't know about yet. I don't know if that is something that would affect these shape outputs, inevitably I need batch_shape `(4)` as you mentioned.

**Comment by csuter** [comment]
Sorry for the half-baked code 😂 I will take a closer look and try actually running it myself when I am back at the computer. 

Meanwhile:

What is the shape of your length scales arg? I believe it will need to be [4, 3]. Also amplitude need only be of shape [4] the trailing one is not needed. I'm also confused that the error says the input is [3, 1, 264] -- did you by chance transpose the gparams arg?

**Comment by mileslucas** [comment]
Hello, I'm actually reopening because it seems this frankenKernel is not positive semi-definite. 

The below code shows that the eigenvalues of the matrix are not all > 0 
```python
amplitude = 10. * tf.ones(4)
length_scales = 10 * tf.ones((4, 3))
se_kernel = tfk.ExponentiatedQuadratic(amplitude)
kernel = InputScaledKernel(se_kernel, length_scales)

mat = kernel.matrix(X, X)
eigs = tf.linalg.eigvalsh(mat)
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    X_, mat_, eigs_ = sess.run([X, mat, eigs])
print(np.all(eigs_ > 0)) # -> False
```
Where X is my input for observation points. It is shape (264, 3) and [here is a data file](https://www.dropbox.com/s/at959cqh6n6qgvi/X.txt?dl=0) for your use. 

This is at the limit of my knowledge of math- I don't have any idea how the above could cause negative eigenvalues. Some short inspection shows the eigenvalues that are negative are very small (~1e-5) but I don't think that would be because of floating point errors. 
```
array([-1.42706864e-04, -1.42597157e-04, -1.28274638e-04, -1.23582911e-04,
       -1.19631666e-04, -1.17399788e-04, -1.15741124e-04, -1.12108442e-04,
       -1.11851608e-04, -1.11851608e-04, -1.11851608e-04, -1.11851608e-04,
       -1.11041219e-04, -1.10283254e-04, -1.09586843e-04, -1.09202432e-04,
       -1.08803863e-04, -1.07279069e-04, -1.05998486e-04, -1.05643965e-04,
       -1.05643965e-04, -1.05643965e-04, -1.03310056e-04, -7.12367182e-05,
       -6.86722124e-05, -6.86722124e-05, -6.86722124e-05, -6.39873833e-05,
       -6.10788557e-05, -6.10788557e-05, -6.10788557e-05, -6.10788557e-05,
       -6.03977060e-05, -5.92993711e-05, -4.94429878e-05, -4.60656629e-05,
       -4.20490906e-05, -3.96636933e-05, -3.86992906e-05, -3.69840600e-05,
       -3.61811981e-05, -3.47077148e-05, -2.61147416e-05, -2.57777228e-05,
       -2.08000365e-05, -1.98761663e-05, -1.69272444e-05, -1.58637667e-05,
       -1.34095426e-05, -1.24312246e-05, -1.20304785e-05, -1.11320005e-05,
       -9.44475687e-06, -7.90955710e-06, -7.79124730e-06, -7.64674951e-06,
       -6.98674512e-06, -6.44641568e-06, -5.91478329e-06, -5.91478329e-06,
       -5.91478329e-06, -5.80676578e-06, -5.79716379e-06, -5.79716379e-06,
       -5.79716379e-06, -5.79716379e-06, -5.45651119e-06, -4.46130980e-06,
       -4.09118547e-06, -3.94376002e-06, -2.92206573e-06, -2.73119304e-06,
       -2.28368481e-06, -1.05545826e-06, -1.01749890e-06, -1.01749890e-06,
       -1.01749890e-06, -1.01749890e-06, -5.48021205e-07], dtype=float32)
```
This is causing the Cholesky decomposition to fail in the creation of the GP. It didn't seem to fail when I used a dummy input
```
X = tf.ones((100, 3))
```
but I don't see how any of the values in my dataset would cause failure.

**Comment by csuter** [comment]
Hey @mileslucas, sorry for the continuing troubles! But glad the definitional issues are behind us -- now this sounds like run-of-the-mill GP numerics issues 😊

First thing I would try would be using float64 throughout. This can be a bit annoying because both numpy and TF seem to strongly prefer float32. So you will need to scrub your program for creation of ndarrays and tensors, forcing each to use 64 bit dtypes. Give that a try and let me know if your problems persist. Another thing to try would be fiddling with the default len scale values you've chosen, to see if any values are stable, but it seems unlikely that that's the issue. Also, any chance there are repeated data in the data set?

**Comment by mileslucas** [comment]
I have gone through my code and verified all tensors have `dtype == float_64` but I still get negative eigenvalues.

For my inputs, I sort of have repeated values- My `X` input are sets of 3 parameters, but the entire set is basically the cartesian product of my parameters. For example
```python
X = itertools.product(
    (6000, 6100, 6200),
    (3.5, 4.0, 4.5),
    (-0.5, 0.0, 0.5)
)
```
so `X` has 264 independent rows, but technically there are many repeated values for the individual parameters. If I make a call to `len(np.unique(X_))` I get `29`, but I'm not sure how that evaluates the uniqueness.

**Comment by csuter** [comment]
Can you possibly share a gist that reproduces the error?

**Comment by mileslucas** [comment]
https://gist.github.com/mileslucas/591c59cde7a021f31393136de59830e4

**Comment by csuter** [comment]
Perfect thanks! I'm away from computer but will take a look later. Most likely tomorrow while traveling home (in Europe right now and it's getting late).

**Comment by mileslucas** [comment]
safe travels, thanks again. Make sure to let me know if you have any issues getting Starfish installed...we're doing a lot of refactoring but that branch should be stable enough for loading the PCAGrid.

**Comment by csuter** [comment]
Will do. Starfish sounds cool!

**Comment by mileslucas** [comment]
Ames has been a great home for me these past four years! I tried implementing your changes and failed...but when I copy and pasted them I succeeded...those are the skills I've gained with my computer science degree!

This works great- I will be trying to use this kernel as part of a specialized likelihood method (instead of a generic Gaussian Process) and hopefully, this should port over well. I really appreciate all of the help!
